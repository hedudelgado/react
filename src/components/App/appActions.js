import { HeaderInterface, InvalidSessionErrorCodeMap, InvalidSessionHelper } from 'mega-code';
import {
  getCountries,
  getNewsletterRegions,
  getNewsletterPreferences,
  getProfile,
  updateProfile,
  deleteCustomer,
} from '../../utils/api';

// actions
export const USER_LOGGED_IN_SUCCESSFULLY = 'USER_LOGGED_IN_SUCCESSFULLY';
export const COUNTRIES_RECEIVED = 'COUNTRIES_RECEIVED';
export const PROFILE_RECEIVED = 'PROFILE_RECEIVED';
export const REGIONS_RECEIVED = 'REGIONS_RECEIVED';
export const NEWSLETTER_RECEIVED = 'NEWSLETTER_RECEIVED';
export const SET_EDIT_VIEW = 'SET_EDIT_VIEW';
export const SET_NOTIFICATION_MESSAGE = 'SET_NOTIFICATION_MESSAGE';
export const PROFILE_UPDATED = 'PROFILE_UPDATED';
export const NEWSLETTER_UPDATED = 'NEWSLETTER_UPDATED';
export const RESET_NEWSLETTER_INFO = 'RESET_NEWSLETTER_INFO';
export const RESET_USER_INFO = 'RESET_USER_INFO';
// views
export const DEFAULT_VIEW = undefined;
export const USER_DETAILS_VIEW = 'USER_DETAILS_VIEW';
export const PASSWORD_VIEW = 'PASSWORD_VIEW';
export const PAYMENT_CARD_VIEW = 'PAYMENT_CARD_VIEW';
export const ADD_GUEST_VIEW = 'ADD_GUEST_VIEW';
export const REGULAR_GUESTS_VIEW = 'REGULAR_GUESTS_VIEW';
export const NEWSLETTER_VIEW = 'NEWSLETTER_VIEW';
export const ROOM_REQUIREMENTS_VIEW = 'ROOM_REQUIREMENTS_VIEW';
export const EXTRAS_PREFERENCES_VIEW = 'EXTRAS_PREFERENCES_VIEW';
export const MEMORABLE_WORD_VIEW = 'MEMORABLE_WORD_VIEW';
// errors
export const PROFILE_ERROR = 'PROFILE_ERROR';
export const PASSWORD_ERROR = 'PASSWORD_ERROR';
export const NEWSLETTER_ERROR = 'NEWSLETTER_ERROR';
export const PAYMENT_CARD_ERROR = 'PAYMENT_CARD_ERROR';
export const USER_DETAILS_ERROR = 'USER_DETAILS_ERROR';
export const REGULAR_GUESTS_ERROR = 'REGULAR_GUESTS_ERROR';
export const MEMORABLE_WORD_ERROR = 'MEMORABLE_WORD_ERROR';
export const ROOM_REQUIREMENTS_ERROR = 'ROOM_REQUIREMENTS_ERROR';
export const EXTRAS_PREFERENCES_ERROR = 'EXTRAS_PREFERENCES_ERROR';

const checkForInvalidSession = (error, config, dispatch) => {
  if (InvalidSessionHelper
    .isSessionExpired(error, config)) {
    dispatch({
      type: RESET_USER_INFO,
    });
    dispatch({
      type: RESET_NEWSLETTER_INFO,
    });
    HeaderInterface.userLogOut(HeaderInterface.LOG_OUT_REASONS.SESSION_EXPIRED);
  }
};

export const getSettings = (email, sessionId, business) => (dispatch, getState) => {
  getProfile(email, sessionId, business).then((profile) => {
    dispatch({
      type: PROFILE_RECEIVED,
      profile,
    });
  }).catch((error) => {
    checkForInvalidSession(error, InvalidSessionErrorCodeMap.ACCOUNT.invalidSession, dispatch);
    dispatch({
      type: PROFILE_ERROR,
      error,
    });
  });
  getCountries()
    .then((countries) => {
      dispatch({
        type: COUNTRIES_RECEIVED,
        countries,
      });
    });
  if (getState().application.showEmailPreferences) {
    getNewsletterRegions()
      .then((regions) => {
        dispatch({
          type: REGIONS_RECEIVED,
          regions,
        });
      });

    getNewsletterPreferences(email)
      .then((newsletter) => {
        dispatch({
          type: NEWSLETTER_RECEIVED,
          newsletter,
        });
      })
      .catch(e => dispatch({
        type: NEWSLETTER_ERROR,
        error: e,
      }));
  }
};

export const updateUserProfile = payload => dispatch => updateProfile(payload).catch((error) => {
  checkForInvalidSession(error, InvalidSessionErrorCodeMap.ACCOUNT.invalidSession, dispatch);
  return Promise.reject(error);
});

export const deleteUserProfile = payload => dispatch => deleteCustomer(payload).catch((error) => {
  checkForInvalidSession(error, InvalidSessionErrorCodeMap.DELETE_PROFILE.invalidSession, dispatch);
  return Promise.reject(error);
});

export default {
  getSettings,
};
