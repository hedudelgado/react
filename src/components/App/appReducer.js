import {
  SET_NOTIFICATION_MESSAGE, SET_EDIT_VIEW, DEFAULT_VIEW,
} from './appActions';

const defaultState = {
  editViewName: DEFAULT_VIEW,
};

export default (state = defaultState, action) => {
  switch (action.type) {
    case SET_EDIT_VIEW:
      return {
        ...state,
        editViewName: action.editViewName,
      };
    case SET_NOTIFICATION_MESSAGE:
      return {
        ...state,
        notificationMessage: action.notificationMessage,
      };
    default:
      return state;
  }
};
