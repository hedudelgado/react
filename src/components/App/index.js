export { default } from './App';

export { default as AppActions } from './appActions';

export { default as AppContainer } from './AppContainer';

export { default as modalReducer } from './modalReducer';
