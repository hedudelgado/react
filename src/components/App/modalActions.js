export const MODAL_OPEN = 'MODAL_OPEN';
export const MODAL_CLOSE = 'MODAL_CLOSE';

export const openModal = content => ({
  type: MODAL_OPEN,
  content,
});

export const closeModal = () => ({
  type: MODAL_CLOSE,
});
