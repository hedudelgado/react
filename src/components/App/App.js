import React, { Component } from 'react';
import {
  array,
  object,
  func,
  bool,
  node,
} from 'prop-types';
import {
  PASSWORD_VIEW,
  PAYMENT_CARD_VIEW,
  USER_DETAILS_VIEW,
  ROOM_REQUIREMENTS_VIEW,
  EXTRAS_PREFERENCES_VIEW,
} from './appActions';
import { ComponentTogglerContainer } from '../ComponentToggler';
import UserProfile from '../UserProfile';
import Password from '../Password';
import UserProfileFormContainer from '../UserProfileForm';
import PasswordFormContainer from '../PasswordForm';
import { PaymentCardContainer } from '../PaymentCard';
import { RegularGuestsContainer } from '../RegularGuests';
import { NewsletterContainer } from '../Newsletter';
import Modal from '../../layout/Modal';
import PaymentCardFormContainer from '../PaymentCardForm';
import { DeleteCustomerContainer } from '../DeleteCustomer';
import RoomRequirements from '../RoomRequirements';
import RoomRequirementsFormContainer from '../RoomRequirementsForm';
import ExtrasPreferences from '../ExtrasPreferences';
import ExtrasPreferencesFormContainer from '../ExtrasPreferencesForm';
import BaseStyle from '../../layout/BaseStyle';
import MemorableWord from '../MemorableWord';
import MoveAccount from '../MoveAccount';
import {
  STAYER, SELF, managePermissionFlag, USER_LOGGED_IN_EVENT,
} from '../../config.json';
import ManagePermission from '../ManagePermission';

class App extends Component {
  constructor(props) {
    super(props);
    this.handleLoggedIn = this.handleLoggedIn.bind(this);
  }

  componentDidMount() {
    window.addEventListener('message', this.handleLoggedIn);
  }

  componentWillUnmount() {
    window.removeEventListener('message', this.handleLoggedIn);
  }

  async handleLoggedIn(event) {
    if (event.data.action === USER_LOGGED_IN_EVENT) {
      const { getSettings } = this.props;
      const { contactDetail: { email }, sessionId, business } = JSON.parse(event.data.profile);
      getSettings(email, sessionId, !!business);
    }
  }

  render() {
    const {
      countries,
      dictionary,
      profile,
      regions,
      newsletter,
      displayModal,
      handleModalClose,
      modalContent,
      userDataReceived,
      application: {
        showEmailPreferences,
      },
    } = this.props;

    const {
      business,
      business: {
        accessLevel = '',
        tetheredGuid = '',
      } = {},
    } = profile;

    const appReady = userDataReceived
      && (!showEmailPreferences || !!regions.length)
      && !!countries.length
      && (!showEmailPreferences || !!Object.keys(newsletter).length)
      && !!Object.keys(dictionary).length
      && !!Object.keys(profile).length;

    const showExtrasPreferences = !(accessLevel === SELF || accessLevel === STAYER);

    return (
      <div data-test="AccountSettings">
        {appReady
        && (
          <BaseStyle className="wb-typography app-container">
            <ComponentTogglerContainer
              viewName={USER_DETAILS_VIEW}
              view={UserProfile}
              edit={UserProfileFormContainer}
            />
            <ComponentTogglerContainer
              viewName={PASSWORD_VIEW}
              view={Password}
              edit={PasswordFormContainer}
            />
            <ComponentTogglerContainer
              viewName={PAYMENT_CARD_VIEW}
              view={PaymentCardContainer}
              edit={PaymentCardFormContainer}
            />
            {managePermissionFlag && (
              <ManagePermission dictionary={dictionary} />
            )}
            <RegularGuestsContainer />
            {showEmailPreferences && (
              <NewsletterContainer />
            )}
            <ComponentTogglerContainer
              viewName={ROOM_REQUIREMENTS_VIEW}
              view={RoomRequirements}
              edit={RoomRequirementsFormContainer}
            />
            {showExtrasPreferences && (
              <ComponentTogglerContainer
                viewName={EXTRAS_PREFERENCES_VIEW}
                view={ExtrasPreferences}
                edit={ExtrasPreferencesFormContainer}
              />
            )}
            {!!tetheredGuid && (
              <MemorableWord />
            )}
            {business && !tetheredGuid && (
              <MoveAccount />
            )}
            {!business && (
            <DeleteCustomerContainer />
            )}
            {displayModal && <Modal data-test="Modal" display content={modalContent} handleClose={handleModalClose} />}
          </BaseStyle>
        )
      }
      </div>
    );
  }
}

App.propTypes = {
  countries: array.isRequired,
  dictionary: object.isRequired,
  profile: object.isRequired,
  getSettings: func.isRequired,
  displayModal: bool.isRequired,
  handleModalClose: func.isRequired,
  modalContent: node,
  newsletter: object.isRequired,
  regions: array.isRequired,
  userDataReceived: bool.isRequired,
  application: object.isRequired,
};

export default App;
