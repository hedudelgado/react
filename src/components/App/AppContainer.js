import { connect } from 'react-redux';
import { getSettings as getSettingsAction } from './appActions';
import { closeModal } from './modalActions';

import App from './App';

const mapStateToProps = ({
  modal: {
    display: displayModal,
    content: modalContent,
  },
  dictionary,
  config,
  countries,
  regions,
  newsletter,
  application,
  userProfile: {
    userDataReceived,
    profile,
  },
}) => ({
  profile,
  displayModal,
  modalContent,
  dictionary,
  config,
  countries,
  regions,
  newsletter,
  application,
  userDataReceived,
});

const mapDispatchToProps = dispatch => ({
  getSettings:
    (email, sessionId, business) => dispatch(getSettingsAction(email, sessionId, business)),
  handleModalClose: () => dispatch(closeModal()),
});

export default connect(mapStateToProps, mapDispatchToProps)(App);
