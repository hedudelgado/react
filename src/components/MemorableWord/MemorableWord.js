import React, { Component } from 'react';
import { object, func } from 'prop-types';
import Typography from '../../layout/Typography';
import Button from '../../layout/Button';
import BaseStyle from '../../layout/BaseStyle';
import Input from '../../layout/Input';
import { validateMemorableWord } from '../../utils/validations';
import { tetherLogin, resetMemorableWord } from '../../utils/api';
import { MEMORABLE_WORD_ERROR, MEMORABLE_WORD_VIEW } from '../App/appActions';
import Notification from '../../layout/Notification';

class MemorableWord extends Component {
  constructor(props) {
    super(props);
    this.state = {
      error: undefined,
      touched: false,
      memorableWord: '',
    };

    this.handleBlur = this.handleBlur.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleSubmit(e) {
    const { profile: { business: { tetheredGuid: guid } }, setNotificationMessage } = this.props;
    const { memorableWord } = this.state;
    e.preventDefault();
    tetherLogin(guid).then(({
      sessionId: worldlineSession,
      sharedSecret: secret,
    }) => resetMemorableWord({
      worldlineSession,
      secret,
      memorableWord,
    }).then(() => {
      setNotificationMessage(MEMORABLE_WORD_VIEW);
      this.setState({ memorableWord: '' });
    })).catch(() => {
      setNotificationMessage(MEMORABLE_WORD_ERROR);
    });
  }

  handleBlur({ target: { value: memorableWord } }) {
    this.setState({
      error: validateMemorableWord(memorableWord),
      touched: true,
    });
  }

  handleChange({ target: { value: memorableWord } }) {
    this.setState({
      error: validateMemorableWord(memorableWord),
      memorableWord,
    });
  }

  render() {
    const { dictionary, app: { notificationMessage } } = this.props;
    const { error, touched, memorableWord } = this.state;
    return (
      <BaseStyle component="section" className="component-container">
        {notificationMessage === MEMORABLE_WORD_ERROR && (
          <Notification data-test="error-notification" priority={2} variant="error" ariaLive="polite" show>
            {dictionary['generic.error']}
          </Notification>
        )}
        {notificationMessage === MEMORABLE_WORD_VIEW && (
          <Notification data-test="success-message" priority={2} variant="success" ariaLive="polite" show>
            {dictionary['update.success.message']}
          </Notification>
        )}
        <Typography component="h4" className="pi-color-greyscale-darkest">
          {dictionary['memorable.word.title']}
        </Typography>
        <Typography component="p" mb="3">{dictionary['memorable.word.body']}</Typography>
        <form name="memorable-word-form" onSubmit={this.handleSubmit}>
          <Input
            onChange={this.handleChange}
            onBlur={this.handleBlur}
            error={dictionary[error]}
            touched={touched}
            type="password"
            name="memorableWord"
            value={memorableWord}
            placeholder={dictionary['memorable.word.placeholder']}
          />
          <Button data-test="update-memorable-word" color="secondary" mt="2" type="submit" fullWidth>
            {dictionary['memorable.word.submit']}
          </Button>
        </form>
      </BaseStyle>
    );
  }
}

MemorableWord.propTypes = {
  dictionary: object.isRequired,
  profile: object.isRequired,
  app: object.isRequired,
  setNotificationMessage: func.isRequired,
};

export default MemorableWord;
