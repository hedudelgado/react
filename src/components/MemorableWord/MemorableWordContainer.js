import { connect } from 'react-redux';
import MemorableWord from './MemorableWord';
import { SET_NOTIFICATION_MESSAGE } from '../App/appActions';

const mapStateToProps = ({
  dictionary,
  userProfile: { profile },
  app,
}) => ({
  dictionary,
  profile,
  app,
});

const mapDispatchToProps = dispatch => ({
  setNotificationMessage: messageName => dispatch({
    type: SET_NOTIFICATION_MESSAGE,
    notificationMessage: messageName,
  }),
});

export default connect(mapStateToProps, mapDispatchToProps)(MemorableWord);
