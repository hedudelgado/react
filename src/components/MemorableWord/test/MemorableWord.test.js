import React from 'react';
import { shallow } from 'enzyme';
import MemorableWord from '../MemorableWord';
import { resetMemorableWord, tetherLogin } from '../../../utils/api';
import { MEMORABLE_WORD_ERROR, MEMORABLE_WORD_VIEW } from '../../App/appActions';

jest.mock('../../../utils/api', () => ({
  __esModule: true,
  resetMemorableWord: jest.fn(() => Promise.resolve({})),
  tetherLogin: jest.fn(() => Promise.resolve({})),
}));

describe('MemorableWord Component', () => {
  let wrapper;
  let props;
  beforeEach(() => {
    props = {
      dictionary: {
        'memorable.word.title': 'Reset your memorable word',
        'memorable.word.body': 'Your memorable word is different from your password and can be used to authorise \'Card not present\' payments.',
        'memorable.word.placeholder': 'New memorable word',
        'memorable.word.error': 'This should be a minimum of 8 characters containing at least one number and capital letter and no special characters',
        'memorable.word.submit': 'Submit',
      },
      profile: {
        business: {
          tetheredGuid: 'guid',
        },
      },
      app: {},
      setNotificationMessage: jest.fn(),
    };
    wrapper = shallow(<MemorableWord {...props} />);
  });

  it('should render the component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should not set touched when user is typing for the first time', () => {
    expect(wrapper.find('[name="memorableWord"]')).toHaveProp('touched', false);
    wrapper.find('[name="memorableWord"]').simulate('change', { target: { value: '' } });
    expect(wrapper.find('[name="memorableWord"]'))
      .toHaveProp('touched', false)
      .toHaveProp('error', props.dictionary['memorable.word.error']);
  });

  it('should set field to touched after blur event', () => {
    expect(wrapper.find('[name="memorableWord"]')).toHaveProp('touched', false);
    wrapper.find('[name="memorableWord"]').simulate('blur', { target: { value: '' } });
    expect(wrapper.find('[name="memorableWord"]'))
      .toHaveProp('touched', true)
      .toHaveProp('error', props.dictionary['memorable.word.error']);
  });

  it('should show success notification', () => {
    props.app.notificationMessage = MEMORABLE_WORD_VIEW;
    wrapper = shallow(<MemorableWord {...props} />);
    expect(wrapper).toContainExactlyOneMatchingElement('[data-test="success-message"]');
  });

  it('should show error notification', () => {
    props.app.notificationMessage = MEMORABLE_WORD_ERROR;
    wrapper = shallow(<MemorableWord {...props} />);
    expect(wrapper).toContainExactlyOneMatchingElement('[data-test="error-notification"]');
  });

  it('should submit form with update success procedure', async () => {
    wrapper.find('[name="memorable-word-form"]').simulate('submit', { preventDefault: () => {} });

    await Promise.all([resetMemorableWord, tetherLogin]).then(() => {
      expect(props.setNotificationMessage)
        .toHaveBeenCalledTimes(1)
        .toHaveBeenCalledWith(MEMORABLE_WORD_VIEW);
    });
  });

  it('should set notification error when tetherLogin fails', async () => {
    tetherLogin.mockImplementationOnce(jest.fn(() => Promise.reject({})));
    wrapper.find('[name="memorable-word-form"]').simulate('submit', { preventDefault: () => {} });

    await Promise.all([tetherLogin]).then(() => {
      expect(props.setNotificationMessage)
        .toHaveBeenCalledTimes(1)
        .toHaveBeenCalledWith(MEMORABLE_WORD_ERROR);
    });
  });
});
