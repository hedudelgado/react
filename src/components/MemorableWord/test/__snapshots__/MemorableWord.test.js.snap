// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MemorableWord Component should render the component 1`] = `
<BaseStyle
  className="component-container"
  component="section"
>
  <Typography
    className="pi-color-greyscale-darkest"
    component="h4"
  >
    Reset your memorable word
  </Typography>
  <Typography
    component="p"
    mb="3"
  >
    Your memorable word is different from your password and can be used to authorise 'Card not present' payments.
  </Typography>
  <form
    name="memorable-word-form"
    onSubmit={[Function]}
  >
    <Input
      mb="wb-push-middle--bottom"
      name="memorableWord"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="New memorable word"
      sessioncamhide={true}
      touched={false}
      type="password"
      value=""
    />
    <Button
      color="secondary"
      data-test="update-memorable-word"
      fullWidth={true}
      href={null}
      mt="2"
      onClick={[Function]}
      submit={false}
      type="submit"
      variant="button"
    >
      Submit
    </Button>
  </form>
</BaseStyle>
`;
